### ISSUE TITLE
GPL-840 Lighthouse_sample data, inconsistency between date_test_string and date_test

### ISSUE THREAD

**Issue by neilsycamore** [sanger/crawler#239]
PAM (Ian W) have noticed data inconsistency in mlwh lighthouse_sample table where there is a date_test_string entry but the date_tested is null. Looking at the examples (877)  the date_test_string in the plate maps has a whitespace!

In addition (but date related) date_test_strings with format '19/07/2020 21:41' result in NULL date_test entries

RT: https://rt.sanger.ac.uk/Ticket/Display.html?id=710743

PAM's SQL
```sql
select * from lighthouse_sample 
where date_tested_string IS NOT NULL AND date_tested IS NULL 
order by created_at desc;
```
=> 877 records (+ve, -ve, void results)

Slack conversation with @pjvv

- The dates format is one issue which we need a feature and data fix for
- The AP ones from May last year look to be a bug in the code which has been fixed so instead of setting the value to NULL is was set to an empty string “”. This could be fixed with a patch.
- The newer ones sometimes have double space between date and time and then space at the end so we need to clean these up before parsing them.

**So, in summary:**

- [x] We need to add a new datetime format for potential dates from the labs
- ~~Data patch to set all date_tested_string to NULL where they are an empty string~~
- [ ] Remove date_tested_string
- [x] Clean the date tested string from the labs with regex before attempting parse

**Comment by pjvv** [comment]
Mongo script to convert string date to BSON date:
```
updateResult = db.samples.update(
  {
    "Date Tested": { $exists: true, $type: "string", $ne: "" },
  },
  [
    {
      $set: {
        "extracted_date": {
          $trim: {
            input: {
              $substrBytes: ["$Date Tested", 0, 20],
            }
          }
        },
      },
    },
    {
      $set: {
        converted_date: {
          $dateFromString: {
            dateString: "$extracted_date",
            format: "%Y-%m-%d %H:%M:%S",
            timezone: "UTC",
            onError: {
              $dateFromString: {
                dateString: "$extracted_date",
                format: "%Y-%m-%d  %H:%M:%S",
                timezone: "UTC",
                onError: {
                  $dateFromString: {
                    dateString: "$extracted_date",
                    format: "%d/%m/%Y %H:%M",
                    timezone: "UTC",
                    onError: {
                      $dateFromString: {
                        dateString: "$extracted_date",
                        format: "%Y/%m/%d %H:%M",
                        timezone: "UTC",
                        onError: null,
                      }
                    },
                  }
                },
              },
            },
          },
        },
      },
    },
    { $unset: "extracted_date" },
    { $unset: "cleaned_date" }
  ],
  { multi: true }
);

printjson(updateResult);
```

**Comment by pjvv** [comment]
Mongo script to rename converted date:
```
// remove the string "Date Tested" field
updateResult = db.samples.update(
    {
        "Date Tested": {
            $exists: true,
            $type: "string",
            $ne: ""
        },
        converted_date: {
            $exists: true,
            $type: "date"
        }
    },
    [
        { $unset: "Date Tested" },
    ],
    { multi: true }
);

printjson(updateResult);

// rename the "converted_date" field
updateResult = db.samples.updateMany(
    {
        converted_date: {
            $exists: true,
            $type: "date"
        }
    },
    { $rename: { 'converted_date': 'Date Tested' } }
);

printjson(updateResult);
```

**Comment by pjvv** [comment]
Mongo script to remove old randox imports and samples:
```
// imports delete
try {
    result = db.imports.deleteMany(
        {
            csv_file_used: {
                $regex: /^lw-randox-biocentre-box-((\d)|(1\d)|20)-.*$/
            }
        }
    );
} catch (e) {
    print(e);
}


// samples count
countResult = db.samples.count(
    {
        source: "Randox",
        file_name: {
            $regex: /^lw-randox-biocentre-box-((\d)|(1\d)|20)-.*$/
        }
    }
)

// samples delete
try {
    result = db.samples.deleteMany(
        {
            source: "Randox",
            file_name: {
                $regex: /^lw-randox-biocentre-box-((\d)|(1\d)|20)-.*$/
            }
        }
    );
} catch (e) {
    print(e);
}
```
